src/cortex_cli/token_manager.py

```diff
@@ -20,70 +20,156 @@
-            'refresh_token': tokens['refresh_token'],
-            'auth_server_url': auth_server_url
-        })
-
-        try:
-            path_to_tokens_dir.mkdir(parents=True, exist_ok=True)
-            with open(Path(path_to_tokens_file), 'w', encoding='UTF-8') as file:
-                file.write(tokens_json)
-        except OSError as error:
-            print('Error writing tokens file', error)
+        tokens = read_tokens(config['tokens_file'])
+
+        new_tokens, status, sleep_time = refresh_tokens(config, tokens, cycle)
+        if new_tokens is None:
+            break
+
+        write_tokens(config['tokens_file'], config['auth_server_url'], status, **new_tokens)
 
         if single_run:
             break
 
-        human_timestamp = timestamp.strftime('%m/%d/%Y %H:%M:%S')
-        print(f'{human_timestamp}: Tokens refreshed successfully.')
-        time.sleep(timeout)
+        time.sleep(sleep_time)
+
+    print(f'{datetime.now().strftime("%m/%d/%Y %H:%M:%S")}: Token manager stopped')
+
+
+def read_tokens(path_to_tokens_file: str) -> dict:
+    """
+    Read current tokens from the tokens file.
+
+    Args:
+        path_to_tokens_file: path to the tokens file
+
+    Returns:
+        dict containing the tokens
+    """
+    with open(path_to_tokens_file, 'r', encoding='utf-8') as file:
+        tokens = json.load(file)
+    return tokens
+
+
+def refresh_tokens(config: dict, current_tokens: dict, cycle: int) -> Tuple[Optional[dict], bool, int]:
+    """
+    Request new tokens from auth server.
+
+    Args:
+        config: dict containing token manager configuration
+        current_tokens: dict containing the current tokens from the tokens file
+        cycle: refresh cycle length in seconds
+
+    Returns:
+        Tuple[Optional[dict], bool, int] = (tokens, status, sleep_time)
+        tokens: dict containing new tokens or current tokens if auth server could not be connected or
+                None if auth server refused to provide new tokens.
+        status: bool, True if tokens were refreshed successfully, False otherwise
+        sleep_time: time to sleep before next refresh attempt
+    """
+    access_token = current_tokens.get('access_token', '')
+    refresh_token = current_tokens.get('refresh_token', '')
+    try:
+        tokens = refresh_request(config['auth_server_url'], config['realm'], config['client_id'], refresh_token)
+        status = True
+        sleep_time = cycle
+        log_timestamp = datetime.now().strftime('%m/%d/%Y %H:%M:%S')
+        print(f'{log_timestamp}: Tokens refreshed successfully.')
+    except (Timeout, ConnectionError) as ex:
+        # No connection to auth server or auth server did not respond, keep current tokens
+        tokens = {'access_token': access_token, 'refresh_token': refresh_token}
+        status = False
+        if isinstance(ex, ConnectionError):
+            sleep_time = AUTH_REQUESTS_TIMEOUT
+        else:
+            sleep_time = 1
```

Comment:
why 1? :D

Reply:
The idea here is that `ConnectionError` is raised almost immediately, but `Timeout` is raised only after the timeout period (currently 20 secs). These times even out the difference, so that the re-connection attempts come at steady pace. I used 1 sec delay here because 0 sec would be even more silly.